src/arithmetic/canonical.cc

```diff
@@ -29,9 +29,17 @@ struct ComExprEntry {
   inline bool operator<(const ComExprEntry& other) const {
     if (level < other.level) return true;
     if (level > other.level) return false;
+    // compare top operator of entries and sort on that if possible (fast check)
     if (value.type_index() < other.value.type_index()) return true;
     if (value.type_index() > other.value.type_index()) return false;
-    return value.get() < other.value.get();
+    // if none of the above distinguishes the terms, compare the expression tree of the entries.
+    // This is a slower check.
+    int compare_result = Compare(value, other.value);
+    if (compare_result < 0) return true;
+    if (compare_result > 0) return false;
+    // it's a problem if we see identical entries at this point. They should've been merged earlier.
+    LOG(FATAL) << "we should not have identical entries at this point";
```

Comment:
since this is operator< let us not do LOG(FATAL) here, as it violates the definition of comparison operator

Reply:
remove this line

Reply:
I agree it violates the definition but I was intending to catch errors there (no ywo entries should be equal). We could replace it with a warning if you agree with some sort of error reporting.

Reply:
I understand this could be useful in this circumstance. except that it is still a bit weird to do so(as users might use the same function for other applications), without assuming an error can be thrown.